smartcards/card.py

```diff
@@ -133,9 +146,22 @@ def read_long_value(self):
             read_so_far += self.CHUNK_SIZE
             chunk_num += 1
 
-        compressed_content = full_bytes[8 +
-                                        self.short_value_length:total_expected_length]
-        return gzip.decompress(compressed_content)
+        raw_content = full_bytes[40 +
+                                 self.short_value_length:total_expected_length]
+        actual_long_value_hash = hashlib.sha256(raw_content).digest()
+
+        if actual_long_value_hash != self.long_value_hash:
+            logging.error(
+                'mismatched hash while reading long value\nexpected: %s\nfound: %s',
+                self.long_value_hash,
+                actual_long_value_hash
+            )
+            return b''
+
+        try:
+            return gzip.decompress(raw_content)
```

Comment:
it's technically possible that the bytes written at the API level would be something gzipped, and then this would be bad news. Can live with it now, but noting it.

Reply:
Yeah, I'd considered that. We could add another byte that says gzipped or not.